src/nvme/nbft.c

```diff
@@ -0,0 +1,739 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+/*
+ * This file is part of libnvme.
+ * Copyright (c) 2021-2022, Dell Inc. or its subsidiaries.  All Rights Reserved.
+ *
+ * Authors: Stuart Hayes <Stuart_Hayes@Dell.com>
+ *
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <arpa/inet.h>
+#include <ccan/endian/endian.h>
+
+#include "private.h"
+#include "nbft.h"
+#include "log.h"
+
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+
+static __u8 csum(void *buffer, int length)
+{
+	int n;
+	__u8 sum = 0;
+
+	for (n = 0; n < length; n++)
+		sum = (__u8)(sum + ((__u8 *)buffer)[n]);
+	return sum;
+}
+
+static void format_ip_addr(char *buf, size_t buflen, __u8 *addr)
+{
+	struct in6_addr *addr_ipv6;
+
+	addr_ipv6 = (struct in6_addr *)addr;
+	if (addr_ipv6->s6_addr32[0] == 0 &&
+	    addr_ipv6->s6_addr32[1] == 0 &&
+	    ntohl(addr_ipv6->s6_addr32[2]) == 0xffff)
+		/* ipv4 */
+		inet_ntop(AF_INET, &(addr_ipv6->s6_addr32[3]), buf, buflen);
+	else
+		/* ipv6 */
+		inet_ntop(AF_INET6, addr_ipv6, buf, buflen);
+}
+
+static bool in_heap(struct nbft_header *header, struct nbft_heap_obj obj)
+{
+	if (le16_to_cpu(obj.length) == 0)
+		return true;
+	if (le32_to_cpu(obj.offset) < le32_to_cpu(header->heap_offset))
+		return false;
+	if (le32_to_cpu(obj.offset) >
+	    le32_to_cpu(header->heap_offset) + le32_to_cpu(header->heap_length))
+		return false;
+	if (le32_to_cpu(obj.offset) + le16_to_cpu(obj.length) >
+	    le32_to_cpu(header->heap_offset) + le32_to_cpu(header->heap_length))
+		return false;
+	return true;
+}
+
+/*
+ *  Return transport_type string (NBFT Table 2)
+ */
+static char *trtype_to_string(__u8 transport_type)
+{
+	switch (transport_type) {
+	case 3:
+		return "tcp";
+	default:
+		return "invalid";
+	}
+}
+
+#define verify(condition, message)					\
+	do {								\
+		if (!(condition)) {					\
+			nvme_msg(NULL, LOG_DEBUG, "file %s: " message "\n", \
```

Comment:
I am not really happy with printing on stdout in the library. In this case it's just for debugging. The fabrics parts of the library we avoided this problem by attaching a logging fd to the `nvme_root_t` object. Obviously, here it's a bit difficult as there is no context we could use to pass in a fd.

Reply:
I agree. But it will take some work to get rid of this one.  Even if I fix this here, there are many more places where nvme_msg is called.  Maybe we can open an issue for this and move forward. 

```
Functions calling this function: nvme_msg

  File   Function          Line
0 nbft.c verify             84 nvme_msg(NULL, LOG_DEBUG, "file %s: " message "\n", \
1 nbft.c __get_heap_obj     99 nvme_msg(NULL, LOG_DEBUG,
2 nbft.c __get_heap_obj    110 nvme_msg(NULL, LOG_DEBUG,
3 nbft.c __get_heap_obj    117 nvme_msg(NULL, LOG_DEBUG,
4 nbft.c read_ssns         231 nvme_msg(NULL, LOG_DEBUG,
5 nbft.c read_ssns         264 nvme_msg(NULL, LOG_DEBUG,
6 nbft.c read_ssns         282 nvme_msg(NULL, LOG_DEBUG,
7 nbft.c read_ssns         291 nvme_msg(NULL, LOG_DEBUG,
8 nbft.c read_hfi_info_tcp 335 nvme_msg(NULL, LOG_DEBUG,
9 nbft.c read_hfi          412 nvme_msg(NULL, LOG_DEBUG,
a nbft.c read_discovery    454 nvme_msg(NULL, LOG_DEBUG,
b nbft.c read_discovery    460 nvme_msg(NULL, LOG_DEBUG,
c nbft.c nvme_nbft_read    695 nvme_msg(NULL, LOG_ERR, "Failed to open %s: %s\n",
d nbft.c nvme_nbft_read    703 nvme_msg(NULL, LOG_ERR, "Failed to read from %s: %s\n",
e nbft.c nvme_nbft_read    715 nvme_msg(NULL, LOG_ERR, "Failed to allocate memory for NBFT table");
f nbft.c nvme_nbft_read    723 nvme_msg(NULL, LOG_ERR, "Failed to read from %s: %s\n",
g nbft.c nvme_nbft_read    737 nvme_msg(NULL, LOG_ERR, "Could not allocate memory for NBFT\n");
h nbft.c nvme_nbft_read    748 nvme_msg(NULL, LOG_ERR, "Failed to parse %s\n", filename);

```

Reply:
Yes, let's move forward and address this later.